File: demo-apps/demoapp003-csharp-specflow/tests/screenplay/support/Actor.cs
namespace DemoApp003.Specs.Screenplay.Support;

public sealed class Actor
{
    private readonly Dictionary<Type, IAbility> _abilities = [];
    private readonly Dictionary<string, object?> _memory = [];

    private Actor(string name) => Name = name;

    public string Name { get; }

    public static Actor Named(string name) => new(name);

    public Actor WhoCan(params IAbility[] abilities)
    {
        foreach (var ability in abilities)
        {
            _abilities[ability.GetType()] = ability;
        }

        return this;
    }

    public void AttemptsTo(params ITask[] tasks)
    {
        foreach (var task in tasks)
        {
            task.PerformAs(this);
        }
    }

    public T Answer<T>(IQuestion<T> question) => question.AnsweredBy(this);

    public TAbility AbilityTo<TAbility>()
        where TAbility : class, IAbility
    {
        if (_abilities.TryGetValue(typeof(TAbility), out var ability))
        {
            return (TAbility)ability;
        }

        throw new InvalidOperationException($"{Name} does not have ability {typeof(TAbility).Name}");
    }

    public void Remember(string key, object? value) => _memory[key] = value;

    public T? Recall<T>(string key, T? defaultValue = default)
    {
        if (!_memory.TryGetValue(key, out var value) || value is null)
        {
            return defaultValue;
        }

        return (T)value;
    }

    public void ForgetAll() => _memory.Clear();
}

File: demo-apps/demoapp002-python-pytest/tests/screenplay/support/actor.py
from __future__ import annotations

from typing import Any, TypeVar

AbilityT = TypeVar("AbilityT")


class Actor:
    def __init__(self, name: str):
        self.name = name
        self._abilities: dict[type, object] = {}
        self._memory: dict[str, Any] = {}

    @classmethod
    def named(cls, name: str) -> "Actor":
        return cls(name)

    def who_can(self, *abilities: object) -> "Actor":
        for ability in abilities:
            self._abilities[type(ability)] = ability
        return self

    def ability_to(self, ability_type: type[AbilityT]) -> AbilityT:
        ability = self._abilities.get(ability_type)
        if ability is None:
            raise RuntimeError(f"{self.name} does not have ability {ability_type.__name__}")
        return ability

    def attempts_to(self, *tasks: object) -> None:
        for task in tasks:
            task.perform_as(self)

    def answer(self, question: object) -> Any:
        return question.answered_by(self)

    def remember(self, key: str, value: Any) -> None:
        self._memory[key] = value

    def recall(self, key: str, default: Any = None) -> Any:
        return self._memory.get(key, default)
